### PR TITLE
User proper markdown parser to compute cell folding

### LINKED ISSUE #158440: Markdown section folding interrupted by python comments
### Applies To

- [X] Notebooks (.ipynb files)
- [ ] Interactive Window and\/or Cell Scripts (.py files with \#%% markers)

### Summary

I expected that folding cells per a markdown header level would collapse all of the markdown and code cells under that header, per the VS Code Outline view of the notebook.

In practice though, if there is a markdown cell with a python code block, then the comments in that python code block are considered the next top-level markdown section which interrupts the code folding.

### Steps

1. Make a new `.ipynb` file and set the interpreter
2. Make markdown cell with content: `# Cell 1`
3. Make markdown cell: `Cell 2: Some markdown content`
4. Make code cell:
```python
# cell 3: print a friendly message
print('hello world')
```
5. Make markdown cell:
```markdown
Cell 4: Execute the following:

```python
# print a friendly message
print('hello world')
```<end the code block>
```
6. Collapse the `# Cell 1` cell
7. Note that cells 2 and 3 are collapsed and hidden, but cell 4 is not

### Work around
Add additional `#`'s to the embedded python comment so that its not at the same level of the makrdown section it is under; e.g.,
```markdown
Cell 4: Execute the following:

```python
## print a friendly message
print('hello world')
```<end the code block>
```

### VS Code Version

1.70.0

### Jupyter Extension Version

v2022.7.1102252217

### Jupyter logs

```shell
Not applicable, since there is no output when folding or unfolding a markdown section.
```

### Coding Language and Runtime Version

Python 3.8.13

### Language Extension Version (if applicable)

v2022.12.0

### Anaconda Version (if applicable)

conda 4.13.0

### Running Jupyter locally or remotely?

Local

### PR DESCRIPTION
**Pull request by mjbvz**
Fixes #158440

### CODE REVIEW

**Comment by rebornix** [comment]
Would this help with https://github.com/microsoft/vscode/issues/157713?

File: src/vs/workbench/contrib/notebook/browser/contrib/outline/notebookOutline.ts [hunk 1]
```diff
@@ -452,12 +452,12 @@ export class NotebookCellOutline implements IOutline<OutlineEntry> {
 
 			if (isMarkdown) {
 				const fullContent = cell.getText().substring(0, 10_000);
-				for (const token of marked.lexer(fullContent, { gfm: true })) {
-					if (token.type === 'heading') {
-						hasHeader = true;
-						entries.push(new OutlineEntry(entries.length, token.depth, cell, renderMarkdownAsPlaintext({ value: token.text }).trim(), false, false));
-					}
+				const headers = [...getMarkdownHeadersInCell(fullContent)];
```
Comment (TylerLeonhardt) [inline src/vs/workbench/contrib/notebook/browser/contrib/outline/notebookOutline.ts:455]: Do you need to use the spread and array? Can't you use `for of` with generators?

File: src/vs/workbench/contrib/notebook/browser/viewModel/foldingModel.ts [hunk 2]
```diff
@@ -319,3 +312,14 @@ export function updateFoldingStateAtIndex(foldingModel: FoldingModel, index: num
 	const range = foldingModel.regions.findRange(index + 1);
 	foldingModel.setCollapsed(range, collapsed);
 }
+
+export function* getMarkdownHeadersInCell(cellContent: string): Iterable<{ readonly depth: number; readonly text: string }> {
```
Comment (TylerLeonhardt) [inline src/vs/workbench/contrib/notebook/browser/viewModel/foldingModel.ts:316]: Why use a generator if you are gonna spread the result into an array?

Comment (mjbvz) [reply src/vs/workbench/contrib/notebook/browser/viewModel/foldingModel.ts:315]: Should be used directly in a `for..of` now

**Review by TylerLeonhardt — APPROVED** [review]

**Pull request merged**